### PR TITLE
KAFKA-12909: disable spurious left/outer stream-stream join fix for old JoinWindows API

### PR DESCRIPTION
**Pull request by mjsax**
We changed the behavior of left/outer stream-stream join via KAFKA-10847.
To avoid a breaking change during an upgrade, we need to disable this
fix by default.

We only enable the fix if users opt-in expliclity by changing their
code. We leverage KIP-633 (KAFKA-8613) that offers a new JoinWindows
API with mandatory grace-period to enable the fix.

Call for review @guozhangwang @spena @ableegoldman @izzyacademy

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:79]: This is the new flag. We set it to `false` if the old methods are used, and to `true` for the new methods from KIP-633.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:70]: Need to change this to be able to add an `JoinWindowsInternal` to access the newly added flag.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:101]: Kept the overlap with KIP-633 PR to a minimum. The key is, that this and the method below must set the new flag to `true`.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:136]: Side fix: `before()` resets grace to 24h (not sure why -- seems to be a bug)

same for `after()` below.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java:157]: Easier to pass one parameter instead of 4

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:68]: This was done by the caller before, ie, `KStreamImplJoin` from above. As we only pass one parameter now, we need to do the flip here if necessary.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java]: Side cleanup

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamStreamJoinNode.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamStreamJoinNode.java:39]: unused.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:100]: We evaluate the feature flag twice. This is the first time. So we add the new flag, too.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamStreamJoinNode.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamStreamJoinNode.java:105]: We evaluate the feature flag twice. This is the second time. So we add the new flag, too.

File: streams/src/test/java/org/apache/kafka/streams/integration/StreamStreamJoinIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/StreamStreamJoinIntegrationTest.java:150]: Need to switch to the new API to keep the fix enabled. Otherwise the test would break.

Just set grace to 24h that was the old default to ensure nothing breaks. -- Similar in other tests below.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplValueJoinerWithKeyTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplValueJoinerWithKeyTest.java:87]: Just some cleanup. The actual change is above.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:66]: unnecessary

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:206]: Did some side cleanup... Will highlight actual changes.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:419]: We can pass nothing to do a check for "no result"

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:516]: Switch to new API. (Not all tests in this class need the switch for this fix.Leave it for KIP-633 PR to update the others when the old API get's deprecated.)

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:78]: Switch to new API (similar below for all other tests in this class)

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:100]: I duplicate this test, to verify that the feature flag, as well as the old API disables this fix. Thus, the usage of the old API in this method should not be changes via KIP-633 PR.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:178]: As above.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:873]: unused

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamOuterJoinTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamOuterJoinTest.java:75]: Same as left-join-test above.

Switch to new API for all cases, and duplicate the "disable fix" test using feature flag and old API.

**Review by spena — APPROVED** [review]
Did 2 passes to the code and it looks good to me!
+1

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:81]: How about moving this copy-constructor to JoinWindowsInternal?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:100]: Thinking a bit more to not evaluate the condition twice: since we call `addStateStore` in the logical plan, a.k.a. `StreamStreamJoinNode` first if both conditions are met, could we just blindly try to get the store from `context.getStateStore()` here, which would return `null` if the store was not added during logical plan and hence does not exist anyways. If it is `null` we can still convert it to an `Optional.empty()`.

**Review by guozhangwang — COMMENTED** [review]
I left a comment on the non-testing part to check if we can just keep one source of truth when evaluating the condition. Otherwise, LGTM.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:81]: Not really possible. Multiple issues:
 - we still need a protected constructor in `JoinWindows`
 - we need to make `graceMs` protected
 - blocker: we need to make the _public_ members `beforeMs` and `afterMs` non-final (and we cannot do this...)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:100]: I don't think that checking the condition twice is a real issue? Also, it seem to be better to do the check, because otherwise (if we might have a bug and incorrectly get a `null` store back) we might mask the bug?

**Comment by mjsax** [comment]
Replied to comments and rebased to resolve merge conflicts. \cc @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:100]: I'm thinking exactly the opposite :) if we have a bug which would cause us to create a state store, checking it twice may actually mask the bug: we would end up creating the state store, and then on the second check not getting it, so the behavior is still correct, and it'll be hard for us to discover we are creating state stores unnecessarily.

If we have a bug and do not create state stores when needed, then we would behave in the old way without the fix; the key point here is that, we only have one decision point to make, and either that decision is correct or buggy, we can get it surfaced quickly.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (izzyacademy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:79]: Thanks for adding this note. This summary is very helpful to me in my implementation.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (izzyacademy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:136]: As per the feedback from @ableegoldman the grace period in the old methods need to stay at 24h. It is in the new methods that we get to specify it as zero.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java
Comment (izzyacademy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java:157]: I like this.

**Comment by izzyacademy** [comment]
@mjsax the changes look good to me. I will rebase my branch after your changes get merged into trunk. Thanks for looping me in.

**Review by guozhangwang — APPROVED** [review]

**Comment by guozhangwang** [comment]
Re-triggered unit tests. We can merge after it turns green.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:136]: Sure. Don't see the connection to this change thought? Note, that `before()` and `after()` are non-static methods, and thus, they should not change/set the grace period. Only the static `of(size)` and non-static `grace()` should change grace period.
```
JoinWindow.of(5000).before(30); // of() will set default of 24h, so no need for before() to reset to 24h, it can just inherit it
JoinWindows.of(5000).grace(40).before(30); // this should leave grace at `40` however, without this fix `before()` would re-set grace to the default of 24h what is incorrect
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:100]: After update the code with the `else` a few tests started to fail. The issue is, that for left/outer join we _always_ set the store name (even if the feature is disabled...) -- only for inner join, we get an `Optinal.empty()`. Thus, we can actually not verify the `else` case (ie, we added a store even if we don't need it) at runtime. I guess we need to rely on the added unit tests instead to cover this case.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:100]: Okay I think I know what's the messy part here: we are setting the stores during KStreamJoin which is only at the parsing phase, but not the logical plan generation phase. The key difference is that the latter has access to the user's configuration whereas the former is not. And because of that we have an unclean settings, we should decide e.g. which stores to create only at the logical plan aka StreamStreamJoinNode.

This is a general issue with Streams parsing/logical plan generation, and I will file a ticket for it.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:100]: Filed https://issues.apache.org/jira/browse/KAFKA-12957

**Pull request merged**
